etcd-agent: tidy cleanup before SIGKILL

gyuho opened the pull request:
https://github.com/golang/go/blob/master/src/os/exec_posix.go#L18 shows that
cmd.Process.Kill calls syscall.SIGKILL to the command. But
http://tldp.org/LDP/Bash-Beginners-Guide/html/sect_12_01.html explains
'If you send a SIGKILL to a process, you remove any chance for the process to
do a tidy cleanup and shutdown, which might have unfortunate consequences.'
This sends SIGTERM, SIGINT syscalls to the PID so that the process could
have more time to clean up the resources.

Related to https://github.com/coreos/etcd/issues/4517, https://github.com/coreos/etcd/issues/4517#issuecomment-183999787

heyitsanthony commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,8 +81,23 @@ func (a *Agent) stop() error {
+	}
+	time.Sleep(time.Second)
+
+	if err := a.cmd.Process.Kill(); err != nil {
```
I think you can get rid of all the `Sleep`s. The final SIGKILL, if any, can happen after timing out on the Wait:
```
go func() {
_, err := a.cmd.Process.Wait()
errc <- err
close(errc)
}()
select {
case <-time.After(5 * time.Second):
a.cmd.Process.Kill()
case err := <-errc:
...
}
err <-errc
...
```

gyuho replied:
@heyitsanthony Just found out Wait returns ProcessState to check if the process exited or not. Will retry your way with it. Thanks.

xiang90 commented:
@gyuho etcd can handle sigint. https://github.com/coreos/etcd/blob/master/pkg/osutil/interrupt_unix.go#L48

So probably you only need to do Process.Signal() [https://golang.org/pkg/os/#Process]. If it fails, then try to kill it. 

@heyitsanthony

gyuho commented:
@xiang90 Good point. Will try that. Thanks.

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,8 +81,26 @@ func (a *Agent) stop() error {
-	err := a.cmd.Process.Kill()
-	if err != nil {
+
+	if err := syscall.Kill(a.cmd.Process.Pid, syscall.SIGTERM); err != nil {
```
use process.Signal. and we only need do sigint.

heyitsanthony commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,8 +81,26 @@ func (a *Agent) stop() error {
+		return err
+	}
+
+	if !a.cmd.ProcessState.Exited() {
```
I don't think this is ever set unless `Wait()` is called first.

gyuho commented:
@xiang90 @heyitsanthony Just addressed on the feedback. PTAL. Thanks!

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,18 +81,56 @@ func (a *Agent) stop() error {
 	}
-	err := a.cmd.Process.Kill()
+
+	err := closeCmd(a.cmd, syscall.SIGTERM)
```
let's only do one signal since we know etcd should and will response to it. drop either sigterm or sigint please.

gyuho replied:
Got it.

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,18 +81,56 @@ func (a *Agent) stop() error {
+	}
+
+	err = a.cmd.Process.Kill()
+	if err.Error() == "os: process already finished" {
```
define the error? it is ugly to do direct string compare

gyuho replied:
Ok will do

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,18 +81,56 @@ func (a *Agent) stop() error {
+	syscall.SIGINT:  "signal: interrupt",
+}
+
+func closeCmd(cmd *exec.Cmd, sig syscall.Signal) error {
```
waitCmdExit

gyuho commented:
@xiang90 PTAL. And just tried the same code in my local machine and works well:

```go
func main() {
	cmd := exec.Command("/home/gyuho/go/bin/etcd")
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	err := cmd.Start()
	if err != nil {
		panic(err)
	}

	time.Sleep(3 * time.Second)

	err = closeCmd(cmd, syscall.SIGTERM)
	if err != nil {
		log.Println(err)
	}

	err = cmd.Process.Kill()
	if err.Error() == "os: process already finished" {
		err = nil
	}
	fmt.Println(err)
}

var sigToState = map[syscall.Signal]string{
	syscall.SIGTERM: "signal: terminated",
}

func closeCmd(cmd *exec.Cmd, sig syscall.Signal) error {
	err := cmd.Process.Signal(sig)
	if err != nil {
		return err
	}

	done, errc := make(chan struct{}), make(chan error)
	go func() {
		st, err := cmd.Process.Wait()
		if err != nil {
			errc <- err
		}
		if st.Exited() {
			done <- struct{}{}
		}
		if st.String() == sigToState[syscall.SIGTERM] {
			done <- struct{}{}
		}
	}()

	select {
	case <-done:
		return nil
	case err := <-errc:
		return err
	case <-time.After(5 * time.Second):
		return fmt.Errorf("timed out")
	}
}

```

heyitsanthony commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -75,23 +77,59 @@ func (a *Agent) start(args ...string) error {
+		a.state = stateStopped
+		err = nil
+	}
+	return err
```
It's possible this function will return before the `Wait()` returns; the process could still be holding its tcp ports.

gyuho commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -75,23 +77,59 @@ func (a *Agent) start(args ...string) error {
 	return nil
 }
 
+var alreadyFinished = errors.New("os: process already finished")
```
This is not defined in os package, which just returns `errors.New("os: process already finished")` in it.

xiang90 replied:
move this error to osutil? ErrProcessAlreadyFinished.

gyuho replied:
Ok will move this.

gyuho commented:
@heyitsanthony Right.

> It's possible this function will return before the Wait() returns; the process could still be holding its tcp ports.

Just added `cmd.Process.Kill` also when there's a time out. PTAL. Thanks.

heyitsanthony commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -75,23 +77,60 @@ func (a *Agent) start(args ...string) error {
+		if st.String() == sigToState[sig] {
+			done <- struct{}{}
+		}
+	}()
```
possible `Wait()` will complete but no channel is signalled. Maybe just
```
go func() {
_, err := cmd.Process.Wait()
errc <- err
}
```
would work just as well.

gyuho replied:
Got it. Will fix. Thanks.

heyitsanthony commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -75,23 +77,60 @@ func (a *Agent) start(args ...string) error {
+		return err
+	case <-time.After(5 * time.Second):
+		cmd.Process.Kill()
+		return fmt.Errorf("timed out")
```
this still needs to know the `Wait()` completed-- an `<-errc` will work if the `done` stuff goes away

heyitsanthony replied:
still needs an `<-errc` after the `Kill`

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -75,23 +77,60 @@ func (a *Agent) start(args ...string) error {
+	return err
+}
+
+var sigToState = map[syscall.Signal]string{
```
it seems that we do not need this map anymore since it is only one kind of sig we will send.

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -75,23 +77,60 @@ func (a *Agent) start(args ...string) error {
+	syscall.SIGTERM: "signal: terminated",
+}
+
+func waitCmdExit(cmd *exec.Cmd, sig syscall.Signal) error {
```
remove the sig arg?

gyuho commented:
@xiang90 @heyitsanthony Just addressed. (Removed done channel, removed sig arg) PTAL.

Thanks!

gyuho commented:
Just moved error to `osutil`.

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,18 +82,40 @@ func (a *Agent) stop() error {
+	return err
+}
+
+func waitCmdExit(cmd *exec.Cmd) error {
```
After second thought, probably SigtermAndWait is a better naming?

/cc @heyitsanthony

xiang90 commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,18 +82,40 @@ func (a *Agent) stop() error {
+	err = a.cmd.Process.Kill()
+	if err.Error() == osutil.ErrProcessAlreadyFinished.Error() {
+		a.state = stateStopped
+		err = nil
```
return nil directly instead of assigning error again.

gyuho replied:
OK will fix.

xiang90 commented on pkg/osutil/osutil.go:
```diff
@@ -16,12 +16,15 @@
 	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/coreos/pkg/capnslog"
 )
 
+var ErrProcessAlreadyFinished = errors.New("os: process already finished")
```
add a comment for this. probably talk about it is not exposed by golang.os pkg.

gyuho replied:
OK will add. Thanks.

heyitsanthony commented on tools/functional-tester/etcd-agent/agent.go:
```diff
@@ -80,18 +82,40 @@ func (a *Agent) stop() error {
 	}
-	_, err = a.cmd.Process.Wait()
+
+	err = a.cmd.Process.Kill()
```
still necessary?

gyuho replied:
Will remove.  Not needed. Thanks!

xiang90 commented:
defer to @heyitsanthony

gyuho commented:
@heyitsanthony PTAL.

1. renamed to `sigtermAndWait` as @xiang90 suggested.
2. added `<-errc` so that the channel can communicate with SIGKILL in case of timeout

Thanks!

heyitsanthony commented:
lgtm

gyuho commented:
great. thanks for help. Will merge in green lights.
